Gnosis safe

leckylao opened the pull request.

leckylao committed:
Implement Gnosis Safe

* Adding Gnosis Safe Env into .env.example;
* Adding build/* into gitignore;
* Added npm packages: @gnosis.pm/safe-contracts, axios, ethereumjs-abi, semver
* Added init Gnosis Safe;
* Added gnosisSafe config/, contracts/, transactions/, getWeb3.js and index.js;

leckylao committed:
fixing getProviderInfo to use await;
fixing getThreshold call;

leckylao committed:
Add getSafeAddresses from owner address;

add getOwnersUriFrom to get API uri;
added getSafeAddresses to get safe addresses from API;

leckylao committed:
implemented submitTx;

setting safeAddress on gnosisSafe constructor;
adding gasNew for estimateGas;
fixing getWeb3 to inside the function;

leckylao committed:
comment out multiSend which not using for now

leckylao committed:
using internal.getAddress instead of web3 requestAccounts;

leckylao committed:
move get safe address inside submitTx;

leckylao committed:
changing to accept spells for submitTx;
fixing bug that txObj value got modified by web3 using txObjClone;

leckylao committed:
added missing SAFE_VERSION_FOR_OFFCHAIN_SIGNATURES

thrilok209 commented on .env.example:
```diff
@@ -1,4 +1,9 @@
+
+# Gnosis Safe
+NODE_ENV="production"
+REACT_APP_NETWORK="production"
```
Why do you need `REACT_APP` key? We are not using it anywhere, right?

thrilok209 commented on .gitignore:
```diff
@@ -23,6 +23,7 @@ dev/test.js
 coverage
 public/css/main.css
 .nyc_output/*
+build/*
```
Have to remove this from gitignore file

thrilok209 commented on src/gnosisSafe/config/production.js:
```diff
@@ -0,0 +1,13 @@
+import {
```
We don't have core DSA contracts on Rinkeby testnet. So I don't think this file is required.

thrilok209 commented on package.json:
```diff
@@ -38,5 +38,11 @@
     "webpack-dev-server": "^3.11.0"
+  },
+  "dependencies": {
+    "@gnosis.pm/safe-contracts": "^1.1.1-dev.2",
```
Do we need this package? If we are just using this to get GnosisSafe ABI `import GnosisSafeSol from "@gnosis.pm/safe-contracts/build/contracts/GnosisSafe.json";`, we can directly copy into our ABI folder, right?

leckylao replied:
As it upgradeable so it could change. By using the package we don't need to maintain it.

thrilok209 commented on src/gnosisSafe/contracts/safeContracts.js:
```diff
@@ -0,0 +1,56 @@
+  "0x34CfAC646f301356fAa8B21e94227e3583Fe3F5F";
+export const DEFAULT_FALLBACK_HANDLER_ADDRESS =
+  "0xd5D82B6aDDc9027B22dCA772Aa68D5d74cdBdF44";
+export const SAFE_MASTER_COPY_ADDRESS_V10 =
```
We are not using `SENTINEL_ADDRESS`, `MULTI_SEND_ADDRESS`,  `SAFE_MASTER_COPY_ADDRESS `, `DEFAULT_FALLBACK_HANDLER_ADDRESS `, `SAFE_MASTER_COPY_ADDRESS_V10 ` addresses anywhere, right?

thrilok209 replied:
If so, can we remove it?

thrilok209 commented on src/gnosisSafe/contracts/safeContracts.js:
```diff
@@ -0,0 +1,56 @@
+import { getWeb3 } from "../getWeb3";
+
+/*
+export const getEncodedMultiSendCallData = (txs, web3) => {
```
Is this commented function required? Or Can we remove it?

thrilok209 commented on src/gnosisSafe/getWeb3.js:
```diff
@@ -0,0 +1,109 @@
+import Web3 from "web3";
```
Can we refactor this file a bit more?

leckylao replied:
Not sure exactly, could you elaborate a bit more?

thrilok209 commented on src/gnosisSafe/index.js:
```diff
@@ -0,0 +1,115 @@
+    if (!_d.to) _d.to = this.instance.address;
+    if (!_d.from) _d.from = _addr;
+    if (!_d.origin) _d.origin = this.origin;
+    if(_d.to === _d.origin)
```
Sometimes `_d.to` and `_d.origin` can be same.

thrilok209 commented on src/gnosisSafe/index.js:
```diff
@@ -0,0 +1,115 @@
+      to: txObj.to,
+      valueInWei: txObj.value,
+      txData: txObj.data,
+      notifiedTransaction: TX_NOTIFICATION_TYPES.STANDARD_TX,
```
Do we need this parameter?

thrilok209 commented on src/gnosisSafe/transactions/notifiedTransactions.js:
```diff
@@ -0,0 +1,11 @@
+export const TX_NOTIFICATION_TYPES = {
```
I think we can remove this one also `TX_NOTIFICATION_TYPES` file, as we are not using anywhere currently.

thrilok209 commented on src/index.js:
```diff
@@ -208,6 +212,8 @@ module.exports = class DSA {
     if (!_d.to) _d.to = this.instance.address;
     if (!_d.from) _d.from = _addr;
     if (!_d.origin) _d.origin = this.origin;
+    if(_d.to === _d.origin)
```
I think you have to change it here also.

thrilok209 commented on src/gnosisSafe/getWeb3.js:
```diff
@@ -0,0 +1,109 @@
+  return firstAddress.toLowerCase() === secondAddress.toLowerCase();
+};
+
+export const ETHEREUM_NETWORK = {
```
Are we using this anywhere?

thrilok209 commented on src/gnosisSafe/getWeb3.js:
```diff
@@ -0,0 +1,109 @@
+    : new Web3(window.web3.currentProvider || "ws://localhost:8545");
+
+let web3 = web3ReadOnly;
+export const getWeb3 = () => web3;
```
Can we get this web3 instance from [here](https://github.com/InstaDApp/dsa-sdk/blob/f3d2245489c21deba0afe6e0f562d83ba74960af/src/index.js#L42)?

leckylao replied:
Originally I thought the same. But the problem is many files need to get Web3. And it's hard to pass it the this.web3 object down as other one is just export const function. So this was what Gnosis safe react is using. Or maybe we can change DSA class constructor to use this one as well to make it consistent?

leckylao committed:
removed non using files, variables and methods;
removing _d.to === _d.origin check as Thilok said it could be the same;
putting build and build/dsa.min.js back;
